### PR TITLE
Disconnect verified

### PR DESCRIPTION
**Pull request by husobee**
### Summary

Phase 2 of supporting connecting/disconnecting verified custodial wallets for #712 

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: migrations/0033_multi_custodian_wallet.up.sql [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+alter table wallets add column wallet_custodian_id uuid default null;
+
+--- create an index on the linking_id (which is how we check linking limits)
+create index wallet_custodian_linking_id_idx on wallet_custodian(linking_id);
+--- create an index on the wallet_id
+create index wallet_custodian_wallet_id_idx on wallet_custodian(wallet_id);
+--- check that the custodian text is in our supported custodians
+alter table wallet_custodian add constraint check_custodian check (
+    custodian IN (
+        'brave', 'uphold', 'bitflyer', 'gemini'));
```
Comment (evq) [inline migrations/0033_multi_custodian_wallet.up.sql:25]: why is brave an option here? ios?

File: wallet/datastore.go [hunk 2]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	// get tx
+	tx, noContextTx := ctx.Value(appctx.DatabaseTransactionCTXKey).(*sqlx.Tx)
+	if !noContextTx {
+		sublogger.Debug().
+			Msg("no tx in context")
+		tx, err = createTx(ctx, pg)
+		if err != nil || tx == nil {
+			sublogger.Error().Err(err).
+				Msg("error creating tx")
+			return fmt.Errorf("failed to create tx for GetCustodianLinkByID: %w", err)
```
Comment (evq) [inline wallet/datastore.go:580]: comment out of sync

File: wallet/datastore.go [hunk 3]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	// get tx
+	tx, noContextTx := ctx.Value(appctx.DatabaseTransactionCTXKey).(*sqlx.Tx)
+	if !noContextTx {
+		sublogger.Debug().
+			Msg("no tx in context")
+		tx, err = createTx(ctx, pg)
+		if err != nil || tx == nil {
+			sublogger.Error().Err(err).
+				Msg("error creating tx")
+			return false, fmt.Errorf("failed to create tx for GetCustodianLinkByID: %w", err)
```
Comment (evq) [inline wallet/datastore.go:677]: comment out of sync

File: wallet/datastore.go [hunk 4]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+			// user deposit information is set, and there is no custodian link
+			return false, nil
+		}
+		// no deposit information set, and no custodian link, we can say it IS migrated
+		// because it has never had the old mono-custodian link performed
+	}
+
+	if !noContextTx {
+		// commit
+		err = tx.Commit()
```
Comment (evq) [inline wallet/datastore.go:709]: why do we need to commit the db transaction here?

File: wallet/datastore.go [hunk 5]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	err = tx.Get(cl, stmt, ID)
+	if err != nil {
+		sublogger.Error().Err(err).
+			Msg("failed to get CustodianLink from DB")
+		return nil, fmt.Errorf("failed to get CustodianLink from DB: %w", err)
+	}
+
+	if !noContextTx {
+		// commit
+		err = tx.Commit()
```
Comment (evq) [inline wallet/datastore.go:769]: same Q here

File: wallet/datastore.go [hunk 6]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+		Str("wallet_id", cl.GetWalletIDString()).
+		Str("custodian", cl.Custodian).
+		Str("linking_id", cl.GetLinkingIDString()).
+		Logger()
+
+	sublogger.Debug().
+		Msg("creating linking of wallet custodian")
+
+	// get tx
+	tx, noContextTx := ctx.Value(appctx.DatabaseTransactionCTXKey).(*sqlx.Tx)
```
Comment (evq) [inline wallet/datastore.go:889]: can we abstract this logic out? seems copy pasted between every new function, here it seems to be slightly different ( missing debug log )

File: wallet/datastore.go [hunk 7]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+		if err != nil || tx == nil {
+			sublogger.Error().Err(err).
+				Msg("error creating tx for wallet linking")
+			return fmt.Errorf("failed to create tx for wallet linking: %w", err)
+		}
+		// add tx to ctx for future
+		defer pg.RollbackTx(tx)
+	}
+
+	// TODO: check here that prior provider linking ids match this new custodian link entry for the given provider
```
Comment (evq) [inline wallet/datastore.go:901]: should we delete this?

File: wallet/datastore.go [hunk 8]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+			return fmt.Errorf("failed to commit transaction: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// helper to create a tx
+func createTx(ctx context.Context, pg *Postgres) (tx *sqlx.Tx, err error) {
+	// get or create tx
```
Comment (evq) [inline wallet/datastore.go:1055]: seems like create tx to me?

File: wallet/datastore.go [hunk 9]
```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	logger(ctx).Debug().
+		Msg("no transaction on context")
```
Comment (evq) [inline wallet/datastore.go:1057]: this is duplicated before some of the calls to `createTx`

Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44

File: wallet/datastore.go (hunk 7 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44

File: wallet/datastore.go (hunk 6 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44

File: wallet/datastore.go (hunk 5 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44

File: migrations/0033_multi_custodian_wallet.up.sql (hunk 1 above)
Comment (husobee) [reply migrations/0033_multi_custodian_wallet.up.sql:24]: correct.

File: wallet/datastore.go (hunk 2 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44

File: wallet/datastore.go (hunk 8 above)
Comment (husobee) [reply wallet/datastore.go]: fixed comment here: https://github.com/brave-intl/bat-go/pull/793/commits/ed9f86432f421ab63e089e309ce1c176e1cbd9a1

File: wallet/datastore.go [hunk 10]
```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	err = tx.Get(&used, stmt, linkingID)
 	if err != nil {
-		return errorutils.Wrap(err, "error looking up used linkings for wallet")
+		sublogger.Error().Err(err).
+			Msg("failed to get CustodianLinkCount from DB")
+		return 0, 0, fmt.Errorf("failed to get CustodianLinkCount from DB: %w", err)
+	}
+	// get max
+	stmt = `
+		select ($2 + count(1)) as max from linking_limit_adjust where provider_linking_id = $1
```
Comment (evq) [inline wallet/datastore.go:457]: previously we used `txGetMaxLinkingSlots` for this, now we have it inlined but that function still exists. should we just call that instead?

File: wallet/datastore.go [hunk 11]
```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
-		return errorutils.Wrap(err, "error looking up max linkings for wallet")
+		return 0, 0, fmt.Errorf("failed to create db transaction GetCustodianLinkByWalletID: %w", err)
 	}
+	// will rollback if tx created at this scope
+	defer rollback()
 
-	usedLinkings, err := txGetUsedLinkingSlots(ctx, tx, providerLinkingID.String())
+	// get used
+	stmt := `
+		select count(distinct(wallet_id)) as used from wallet_custodian where linking_id = $1
```
Comment (evq) [inline wallet/datastore.go:447]: should we update the logic in `txGetUsedLinkingSlots` to use this query? ( and call `txGetUsedLinkingSlots` )

File: wallet/datastore.go [hunk 12]
```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		sublogger.Error().Err(err).
+			Msg("failed to insert wallet_custodian")
+		return fmt.Errorf("failed to insert wallet custodian record: %w", err)
+	}
+
+	if !uuid.Equal(existingLinkingID, *new(uuid.UUID)) {
+		// check if the member matches the associated member
+		if !uuid.Equal(*cl.LinkingID, existingLinkingID) {
+			return handlers.WrapError(errors.New("wallets do not match"), "unable to match wallets", http.StatusForbidden)
```
Comment (evq) [inline wallet/datastore.go:673]: does the client handle this case?

File: wallet/datastore.go [hunk 13]
```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
...
+	err = tx.Get(cl, stmt, cl.WalletID, cl.Custodian, cl.LinkingID)
 	if err != nil {
-		return err
+		sublogger.Error().Err(err).
+			Msg("failed to insert wallet_custodian")
+		return fmt.Errorf("failed to insert wallet custodian record: %w", err)
+	}
+	// update wallets with new deposit destination
+	stmt = `
+		update wallets set user_deposit_destination=$1 where id=$2
```
Comment (evq) [inline wallet/datastore.go:723]: if we're already updating wallets why not keep the other linking fields up to date?

**Review by evq — CHANGES_REQUESTED** [review]

File: wallet/datastore.go (hunk 12 above)
Comment (husobee) [reply wallet/datastore.go:673]: This is how we currently respond to this type of error.  I can ask if they appropriately handle, but is is outlined as a known error condition in the spec.

File: wallet/datastore.go (hunk 11 above)
Comment (husobee) [reply wallet/datastore.go:447]: :+1:

File: wallet/datastore.go (hunk 10 above)
Comment (husobee) [reply wallet/datastore.go:457]: :+1:

File: wallet/datastore.go (hunk 13 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/dcbcd7ca13f72be4249e0760d01398e454c5bc66

File: wallet/datastore.go (hunk 11 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/dcbcd7ca13f72be4249e0760d01398e454c5bc66

File: wallet/datastore.go (hunk 10 above)
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/dcbcd7ca13f72be4249e0760d01398e454c5bc66

**Comment by tsmartt** [comment]
@husobee thanks for the walkthrough. Can we add a relinking test for each wallet? I.e. if I create, disconnect, then relink -- on relink the DB record should be updated/reused, not a new one created.

**Review by tsmartt — COMMENTED** [review]
@husobee thanks for the walkthrough. Can we add a relinking test for each wallet? I.e. if I create, disconnect, then relink -- on relink the DB record should be updated/reused, not a new one created.

**Review by tsmartt — DISMISSED** [review]
Great work!

**Review by evq — DISMISSED** [review]

File: wallet/datastore.go [hunk 14]
```diff
@@ -340,6 +263,7 @@ func txGetMaxLinkingSlots(ctx context.Context, tx *sqlx.Tx, providerLinkingID st
 		select ($2 + count(1)) as max from linking_limit_adjust where provider_linking_id = $1
 	`
 	err := tx.Get(&max, statement, providerLinkingID, getEnvMaxCards())
+	fmt.Println("err: ", err)
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/datastore.go]: resolved: https://github.com/brave-intl/bat-go/pull/793/commits/b23ffed8d4634b439a8f37ed3815879b3b3a832e

**Pull request merged**